### PR TITLE
Removed libtiny-ui

### PR DESCRIPTION
**Pull request by trevarj**
Added TinyUI struct to combine the TUI and Logger. Removes dynamic
dispatch from the UI.

This will allow us to access/configure the Logger from commands.

### CODE REVIEW

File: libtiny_common/src/lib.rs [hunk 1]
```diff
@@ -141,3 +141,97 @@ impl Hash for ChanNameRef {
...
+            MsgSource::Chan { chan, .. } => chan.display(),
+            MsgSource::User { nick, .. } => nick,
+        }
+    }
+}
+
+// NOTE: Keep the variants sorted in increasing significance, to avoid updating
+// style with higher significance for a less significant style (e.g. updating
+// from `Highlight` to `NewMsg` in `set_tab_style`).
+// TODO: This shouldn't be a part of the API
```
Comment (trevarj) [inline libtiny_common/src/lib.rs:212]: need to remove this

File: libtiny_tui/src/lib.rs [hunk 2]
```diff
@@ -204,17 +202,17 @@ async fn input_handler<S>(
...
             }
         }
     }
 }
 
-impl UI for TUI {
+impl TUI {
     delegate!(draw());
-    delegate!(new_server_tab(serv_name: &str, alias: Option<String>,));
+    delegate!(new_server_tab(serv_name: &str, alias: &Option<String>,));
```
Comment (osa1) [inline libtiny_tui/src/lib.rs:215]: Why did you have to change this?

File: tiny/src/ui.rs [hunk 3]
```diff
@@ -7,16 +7,78 @@ use crate::cmd::{parse_cmd, CmdArgs, ParseCmdResult};
...
+            self.ui.$name( $( $x, )* );
+            if let Some(logger) = &self.logger {
+                logger.$name( $( $x, )* );
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct TinyUI {
```
Comment (osa1) [inline tiny/src/ui.rs:29]: Let's call this `UI`; it's shorter and there isn't any ambiguity about what's UI is this.

**Review by osa1 — COMMENTED** [review]
Thanks for doing this. Just a very quick look for now, I'll review in more detail later.

File: libtiny_tui/src/lib.rs (hunk 2 above)
Comment (trevarj) [reply libtiny_tui/src/lib.rs:215]: To make this macro happy:
`ui.rs`
```
macro_rules! delegate {
    ( $name:ident ( $( $x:ident: $t:ty, )* )) => {
        pub(crate) fn $name(&self, $($x: $t,)*) {
            self.ui.$name( $( $x, )* );
            if let Some(logger) = &self.logger {
                logger.$name( $( $x, )* );
            }
        }
    }
}
```

Let me know if you can spot a better way. I'm a macro noob

File: libtiny_tui/src/tui.rs [hunk 4]
```diff
@@ -371,7 +370,7 @@ impl TUI {
     }
 
     /// Returns index of the new tab if a new tab is created.
-    pub fn new_server_tab(&mut self, serv: &str, alias: Option<String>) -> Option<usize> {
+    pub fn new_server_tab(&mut self, serv: &str, alias: &Option<String>) -> Option<usize> {
```
Comment (osa1) [inline libtiny_tui/src/tui.rs:373]: If you revert this change (and the `clone()` below) you won't have to make the other changed I asked about in the previous review.

File: libtiny_tui/src/lib.rs (hunk 2 above)
Comment (osa1) [reply libtiny_tui/src/lib.rs:215]: No need to change the macro, I added another comment below.

File: tiny/src/main.rs [hunk 5]
```diff
@@ -127,20 +127,17 @@ fn run(
...
+                    logger.new_server_tab("mentions", &None);
                     Some(logger)
                 }
             });
 
-        let tui: Box<dyn UI> = match logger {
-            None => Box::new(tui),
-            Some(logger) => Box::new(libtiny_ui::combine(tui, logger)),
-        };
+        let tui = UI::new(tui, logger);
```
Comment (osa1) [inline tiny/src/main.rs:135]: :+1:

**Review by osa1 — APPROVED** [review]
Looks great! I think we could revert the change to `&Option<String>` from `Option<String>` in some of the methods, then the PR should be much smaller and you can remove clonning in a few places. Also this currently doesn't pass the tests. If you could fix those I'll merge it.

Thanks!

File: libtiny_tui/src/lib.rs (hunk 2 above)
Comment (trevarj) [reply libtiny_tui/src/lib.rs:215]: @osa1 This is the reason why I changed it:
image

File: libtiny_logger/src/lib.rs [hunk 6]
```diff
@@ -37,25 +36,21 @@ impl Logger {
...
             self.inner.borrow_mut().$name( $( $x, )* )
         }
     }
 }
 
-impl UI for Logger {
-    fn draw(&self) {}
-    delegate!(new_server_tab(serv: &str, alias: Option<String>,));
+impl Logger {
+    pub fn draw(&self) {}
```
Comment (osa1) [inline libtiny_logger/src/lib.rs:46]: This could also be removed
